src/libmongoc/src/mongoc/mongoc-client.c

```diff
@@ -1140,6 +1123,7 @@ _mongoc_client_new_from_uri (mongoc_topology_t *topology)
    const char *appname;
 
    BSON_ASSERT (topology);
+   BSON_ASSERT (topology->valid);
```

Comment:
I can remove this if there's a concern it may break something; however, I've added `topology->valid` checks in the various functions that would construct a topology before calling this method and I don't believe we'd ever enter this function with an invalid topology.

That said, I'm not sure what the code path is for `mongoc_client_pool_new` constructing a pool with an invalid topology. If the application then tries to pop a client from that pool, is it possible that `_mongoc_client_new_from_topology` (previously `_mongoc_client_new_from_uri`) gets called with an invalid topology?

Reply:
```suggestion
   
   if (!topology->valid) {
      MONGOC_ERROR ("cannot construct client from invalid topology");
      return NULL;
   }
```
In this case, I think a warning + `NULL` result would be preferable to an assertion. If `mongoc_topology_new ()` had guaranteed `NULL` result on invalid topology, then I would have been in favor of this assertion, but alas.